### PR TITLE
fix(langgraph-checkpoint-aws): handle RetryableConflictException gracefully

### LINKED ISSUE #798: Handling `RetryableConflictException` gracefully when creating event
### Package

langgraph-checkpoint-aws

### Checked other resources

- [x] I added a descriptive title to this issue
- [x] I searched the LangChain documentation with the integrated search
- [x] I used the GitHub search to find a similar issue and didn't find it
- [x] I am sure this is a bug and not a question or request for help

### Example Code

```python
We should handle the `RetryableConflictException` when creating events in bedrock-agentcore memory.

I do not have a minimal reproducible example, as this is a general expectation
```

### Error Message and Stack Trace (if applicable)

```shell
"  File ""/opt/amazon/lib/python3.11/site-packages/ec2_dns_resolver_ticket_operator/processor/message_processor.py"", line 146, in _process_ticket_with_comments"
    for message_from_agent in self._ticket_handler.handle_ticket(
"  File ""/opt/amazon/lib/python3.11/site-packages/ec2_dns_resolver_ticket_operator/handler/ticket_handler.py"", line 81, in handle_ticket"
    for event in agent.graph.stream(
"  File ""/opt/amazon/lib/python3.11/site-packages/langgraph/pregel/main.py"", line 2679, in stream"
    for _ in runner.tick(
"  File ""/opt/amazon/lib/python3.11/site-packages/langgraph/pregel/_runner.py"", line 258, in tick"
    _panic_or_proceed(
"  File ""/opt/amazon/lib/python3.11/site-packages/langgraph/pregel/_runner.py"", line 520, in _panic_or_proceed"
    raise exc
"  File ""/opt/amazon/lib/python3.11/site-packages/langgraph/pregel/_executor.py"", line 80, in done"
    task.result()
"  File ""/opt/amazon/python3.11/lib/python3.11/concurrent/futures/_base.py"", line 449, in result"
    return self.__get_result()
           ^^^^^^^^^^^^^^^^^^^
"  File ""/opt/amazon/python3.11/lib/python3.11/concurrent/futures/_base.py"", line 401, in __get_result"
    raise self._exception
"  File ""/opt/amazon/python3.11/lib/python3.11/concurrent/futures/thread.py"", line 58, in run"
"    result = self.fn(*self.args, **self.kwargs)"
             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
"  File ""/opt/amazon/lib/python3.11/site-packages/langgraph/pregel/_retry.py"", line 42, in run_with_retry"
"    return task.proc.invoke(task.input, config)"
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
"  File ""/opt/amazon/lib/python3.11/site-packages/langgraph/_internal/_runnable.py"", line 656, in invoke"
"    input = context.run(step.invoke, input, config, **kwargs)"
            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
"  File ""/opt/amazon/lib/python3.11/site-packages/langgraph/_internal/_runnable.py"", line 400, in invoke"
"    ret = self.func(*args, **kwargs)"
          ^^^^^^^^^^^^^^^^^^^^^^^^^^
"  File ""/opt/amazon/lib/python3.11/site-packages/ec2_dns_resolver_ticket_operator/models/agent/component_agent.py"", line 453, in <lambda>"
    lambda state: populate_autocut_details(
                  ^^^^^^^^^^^^^^^^^^^^^^^^^
"  File ""/opt/amazon/lib/python3.11/site-packages/ec2_dns_resolver_ticket_operator/models/agent/component_agent.py"", line 76, in populate_autocut_details"
    aws_account_agent_response: Optional[AwsAccountAgentResponse] = fetch_aws_account_names(
                                                                    ^^^^^^^^^^^^^^^^^^^^^^^^
"  File ""/opt/amazon/lib/python3.11/site-packages/ec2_dns_resolver_ticket_operator/models/agent/agent_methods.py"", line 181, in fetch_aws_account_names"
"    background_info = fetch_background_info(state, llm, tools, _config)"
                      ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
"  File ""/opt/amazon/lib/python3.11/site-packages/ec2_dns_resolver_ticket_operator/models/agent/agent_methods.py"", line 160, in fetch_background_info"
"    agent_response = agent.invoke({""messages"": prompt_to_fetch_background})"
                     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
"  File ""/opt/amazon/lib/python3.11/site-packages/langgraph/pregel/main.py"", line 3094, in invoke"
    for chunk in self.stream(
"  File ""/opt/amazon/lib/python3.11/site-packages/langgraph/pregel/main.py"", line 2615, in stream"
    with SyncPregelLoop(
"  File ""/opt/amazon/lib/python3.11/site-packages/langgraph/pregel/_loop.py"", line 1100, in __exit__"
"    return self.stack.__exit__(exc_type, exc_value, traceback)"
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
"  File ""/opt/amazon/python3.11/lib/python3.11/contextlib.py"", line 601, in __exit__"
    raise exc_details[1]
"  File ""/opt/amazon/python3.11/lib/python3.11/contextlib.py"", line 586, in __exit__"
    if cb(*exc_details):
       ^^^^^^^^^^^^^^^^
"  File ""/opt/amazon/lib/python3.11/site-packages/langgraph/pregel/_executor.py"", line 117, in __exit__"
    task.result()
"  File ""/opt/amazon/python3.11/lib/python3.11/concurrent/futures/_base.py"", line 449, in result"
    return self.__get_result()
           ^^^^^^^^^^^^^^^^^^^
"  File ""/opt/amazon/python3.11/lib/python3.11/concurrent/futures/_base.py"", line 401, in __get_result"
    raise self._exception
"  File ""/opt/amazon/lib/python3.11/site-packages/langgraph/pregel/_executor.py"", line 80, in done"
    task.result()
"  File ""/opt/amazon/python3.11/lib/python3.11/concurrent/futures/_base.py"", line 449, in result"
    return self.__get_result()
           ^^^^^^^^^^^^^^^^^^^
"  File ""/opt/amazon/python3.11/lib/python3.11/concurrent/futures/_base.py"", line 401, in __get_result"
    raise self._exception
"  File ""/opt/amazon/python3.11/lib/python3.11/concurrent/futures/thread.py"", line 58, in run"
"    result = self.fn(*self.args, **self.kwargs)"
             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
"  File ""/opt/amazon/lib/python3.11/site-packages/langgraph/pregel/_loop.py"", line 996, in _checkpointer_put_after_previous"
    prev.result()
"  File ""/opt/amazon/python3.11/lib/python3.11/concurrent/futures/_base.py"", line 449, in result"
    return self.__get_result()
           ^^^^^^^^^^^^^^^^^^^
"  File ""/opt/amazon/python3.11/lib/python3.11/concurrent/futures/_base.py"", line 401, in __get_result"
    raise self._exception
"  File ""/opt/amazon/lib/python3.11/site-packages/langgraph/pregel/_executor.py"", line 80, in done"
    task.result()
"  File ""/opt/amazon/python3.11/lib/python3.11/concurrent/futures/_base.py"", line 449, in result"
    return self.__get_result()
           ^^^^^^^^^^^^^^^^^^^
"  File ""/opt/amazon/python3.11/lib/python3.11/concurrent/futures/_base.py"", line 401, in __get_result"
    raise self._exception
"  File ""/opt/amazon/python3.11/lib/python3.11/concurrent/futures/thread.py"", line 58, in run"
"    result = self.fn(*self.args, **self.kwargs)"
             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
"  File ""/opt/amazon/lib/python3.11/site-packages/langgraph/pregel/_loop.py"", line 996, in _checkpointer_put_after_previous"
    prev.result()
"  File ""/opt/amazon/python3.11/lib/python3.11/concurrent/futures/_base.py"", line 449, in result"
    return self.__get_result()
           ^^^^^^^^^^^^^^^^^^^
"  File ""/opt/amazon/python3.11/lib/python3.11/concurrent/futures/_base.py"", line 401, in __get_result"
    raise self._exception
"  File ""/opt/amazon/lib/python3.11/site-packages/langgraph/pregel/_executor.py"", line 80, in done"
    task.result()
"  File ""/opt/amazon/python3.11/lib/python3.11/concurrent/futures/_base.py"", line 449, in result"
    return self.__get_result()
           ^^^^^^^^^^^^^^^^^^^
"  File ""/opt/amazon/python3.11/lib/python3.11/concurrent/futures/_base.py"", line 401, in __get_result"
    raise self._exception
"  File ""/opt/amazon/python3.11/lib/python3.11/concurrent/futures/thread.py"", line 58, in run"
"    result = self.fn(*self.args, **self.kwargs)"
             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
"  File ""/opt/amazon/lib/python3.11/site-packages/langgraph/pregel/_loop.py"", line 996, in _checkpointer_put_after_previous"
    prev.result()
"  File ""/opt/amazon/python3.11/lib/python3.11/concurrent/futures/_base.py"", line 456, in result"
    return self.__get_result()
           ^^^^^^^^^^^^^^^^^^^
"  File ""/opt/amazon/python3.11/lib/python3.11/concurrent/futures/_base.py"", line 401, in __get_result"
    raise self._exception
"  File ""/opt/amazon/lib/python3.11/site-packages/langgraph/pregel/_executor.py"", line 80, in done"
    task.result()
"  File ""/opt/amazon/python3.11/lib/python3.11/concurrent/futures/_base.py"", line 449, in result"
    return self.__get_result()
           ^^^^^^^^^^^^^^^^^^^
"  File ""/opt/amazon/python3.11/lib/python3.11/concurrent/futures/_base.py"", line 401, in __get_result"
    raise self._exception
"  File ""/opt/amazon/python3.11/lib/python3.11/concurrent/futures/thread.py"", line 58, in run"
"    result = self.fn(*self.args, **self.kwargs)"
             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
"  File ""/opt/amazon/lib/python3.11/site-packages/langgraph/pregel/_loop.py"", line 998, in _checkpointer_put_after_previous"
"    cast(BaseCheckpointSaver, self.checkpointer).put("
"  File ""/opt/amazon/lib/python3.11/site-packages/langgraph_checkpoint_aws/agentcore/saver.py"", line 203, in put"
    self.checkpoint_event_client.store_blob_events_batch(
"  File ""/opt/amazon/lib/python3.11/site-packages/langgraph_checkpoint_aws/agentcore/helpers.py"", line 178, in store_blob_events_batch"
    self.client.create_event(
"  File ""/opt/amazon/lib/python3.11/site-packages/botocore/client.py"", line 602, in _api_call"
"    return self._make_api_call(operation_name, kwargs)"
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
"  File ""/opt/amazon/lib/python3.11/site-packages/botocore/context.py"", line 123, in wrapper"
"    return func(*args, **kwargs)"
           ^^^^^^^^^^^^^^^^^^^^^
"  File ""/opt/amazon/lib/python3.11/site-packages/botocore/client.py"", line 1078, in _make_api_call"
"    raise error_class(parsed_response, operation_name)"
botocore.exceptions.ClientError: An error occurred (RetryableConflictException) when calling the CreateEvent operation: Concurrently modified resources. Please retry again.
```

### Description

During event creation when interacting with bedrock-agentcore, the library should handle the retryable exception as stated in the [documentation](https://docs.aws.amazon.com/AWSJavaSDK/latest/javadoc/com/amazonaws/services/clouddirectory/model/RetryableConflictException.html)

### PR DESCRIPTION
**Pull request by ps06756**
## Summary

This PR adds automatic retry logic to gracefully handle `RetryableConflictException` when creating events in bedrock-agentcore memory.

## Problem

The `langgraph-checkpoint-aws` package currently does not handle `RetryableConflictException` that occurs during concurrent resource modifications when creating events in AWS Bedrock AgentCore memory. This transient error is designed to be retried according to AWS documentation, but the library was propagating it directly to users.

## Solution

- Created `BedrockAgentCoreClientEnhanced` wrapper class that automatically retries on `RetryableConflictException`
- Implements exponential backoff strategy (starts at 100ms, doubles each retry, max 2 seconds)
- Defaults to 3 retry attempts
- Non-retryable errors fail immediately without retry
- Clean separation of concerns - retry logic is encapsulated in the wrapper, keeping helper methods simple

## Changes

- `langgraph_checkpoint_aws/agentcore/helpers.py`: Added `BedrockAgentCoreClientEnhanced` wrapper class with retry logic
- `tests/unit_tests/agentcore/test_saver.py`: Added unit tests covering various retry scenarios

## Testing

- ✅ All 55 unit tests pass (48 existing + 7 new)
- ✅ Linting passed (ruff)
- ✅ Type checking passed (mypy)

## Checklist

- [x] Tests fail when the new logic is broken
- [x] Happy path is covered
- [x] Edge cases and error conditions are tested
- [x] Uses fixtures/mocks for external dependencies
- [x] Tests are deterministic (no flaky tests)
- [x] No breaking changes to public API

Fixes #798

### CODE REVIEW

**Comment by ps06756** [comment]
@michaelnchin Can you please help review this ?

**Comment by michaelnchin** [comment]
Hi @ps06756! Apologies for the delay - will have a look at this soon.

File: libs/langgraph-checkpoint-aws/langgraph_checkpoint_aws/agentcore/helpers.py [hunk 1]
```diff
@@ -40,6 +42,81 @@
 # Union type for all events
 EventType = CheckpointEvent | ChannelDataEvent | WritesEvent
 
+# Default retry configuration
+DEFAULT_MAX_RETRIES = 3
+DEFAULT_INITIAL_BACKOFF = 0.1  # 100ms
+DEFAULT_MAX_BACKOFF = 2.0  # 2 seconds
+
+
+class BedrockAgentCoreClientEnhanced:
```
Comment (michaelnchin) [inline libs/langgraph-checkpoint-aws/langgraph_checkpoint_aws/agentcore/helpers.py]: This naming feels too vague, let's rename to something explicit like `BedrockAgentCoreClientWithRetry`.

File: libs/langgraph-checkpoint-aws/langgraph_checkpoint_aws/agentcore/helpers.py [hunk 2]
```diff
@@ -40,6 +42,81 @@
 # Union type for all events
 EventType = CheckpointEvent | ChannelDataEvent | WritesEvent
 
+# Default retry configuration
+DEFAULT_MAX_RETRIES = 3
+DEFAULT_INITIAL_BACKOFF = 0.1  # 100ms
+DEFAULT_MAX_BACKOFF = 2.0  # 2 seconds
```
Comment (michaelnchin) [inline libs/langgraph-checkpoint-aws/langgraph_checkpoint_aws/agentcore/helpers.py:48]: Can we make this configuration user configurable instead of hardcoding the values?

File: libs/langgraph-checkpoint-aws/tests/unit_tests/agentcore/test_saver.py [hunk 3]
```diff
@@ -1244,3 +1246,188 @@ def test_build_checkpoint_tuple_no_parent(
...
+            sessionId="test-session",
+            eventTimestamp="2024-01-01T00:00:00Z",
+            payload=[{"blob": "test"}],
+        )
+        elapsed_time = time.time() - start_time
+
+        assert result == {"eventId": "test-event-id"}
+        assert mock_boto_client.create_event.call_count == 3
+        # Should have slept at least 0.1s (first retry) + 0.2s (second retry) = 0.3s
+        assert elapsed_time >= 0.3
```
Comment (michaelnchin) [inline libs/langgraph-checkpoint-aws/tests/unit_tests/agentcore/test_saver.py]: I really think we should mock `time.sleep` here instead of observing the sleep duration on run because:
1) Using real `time.sleep` forces the test to run for the max duration of 0.3s, which will slow down the CI suite as a whole
2) We're not verifying the more fine-graining exponential backoff values when we use a single `assert elapsed_time >= 0.3` statement.

With the mock, the test should look something like:

```python
@patch('time.sleep')
def test_retry_exponential_backoff(self, mock_sleep, enhanced_client, mock_boto_client):
    """Test that exponential backoff is applied between retries."""
    ...
    assert mock_sleep.call_args_list == [
        call(0.1),  # 0.1 * 2^0
        call(0.2),  # 0.1 * 2^1
    ]
```

**Comment by ps06756** [comment]
@michaelnchin I have addressed the above comments

**Comment by michaelnchin** [comment]
Integration tests passing: https://github.com/langchain-ai/langchain-aws/actions/runs/20944082694/job/60183530812

**Review by michaelnchin — APPROVED** [review]
Thank you @ps06756 !
